Change taker's commitments logic

PulpCattel opened the pull request:
Changes taker's logic so that it sources commitments only from the same mixdepth from which we are spending (and externals).

Tests are passing and I've done a couple mixes on regtest to check (including triggering failure).

Are there other places (like error messages or docs) where this should be changed?

AdamISZ commented:
As discussed offline I am with you on this, I think this is somewhere between very preferable and needed. Thanks for taking the initiative. Rather swamped with other things right now, will get back to this soon.

AdamISZ commented:
This is important, I'm kind of surprised it didn't get any attention from anyone else yet - is it OK if PoDLE revelations show utxos to makers that are from different mixdepths? Generally one would think - definitely not, it's a violation of the privacy boundary. It won't happen too often but it's reasonable to say it should never happen at all; the only issue that has to be addressed is, does it lead to an unacceptably high failure rate due to failure to source commitments in, say, a typical tumbler run? I very much hope not, and I think not, but perhaps a little investigation is in order.

PulpCattel commented:
It's a bit weird to me, as an:

> unacceptably high failure rate due to failure to source commitments

would actually mean that this privacy leak happens often.

I.e., if cause to this change many tumbler runs fail, wouldn't they leak a lot otherwise? 
I don't see a scenario where tumbler runs aren't failing, but we are also not leaking too much (I'm talking about average users, of course it is possible to achieve both with careful decision-making; e.g., by filling external commitments with a chosen set of UTXOs that we deem right for the job).

IDK if you guys see it the same way, but I would call this "good" failures, since it is basically a fail-safe.

If during a tumbler run the taker ends up repeatedly in need of sourcing UTXOs from other mixdepths, I think the problem is more deep.
As I said to you, as workaround, maybe there could be an option in the settings (like either `source mixdepth` or `source wallet`) and the user decides.
Even in a case like that, I would prefer a mixdepth only sourcing as default.

This change in taker's logic could also turn `add-utxo.py` script in a more known and used tool, which would make sense no?
Sourcing UTXOs for commitments should be a somewhat more conscious decision.

AdamISZ commented:
> If during a tumbler run the taker ends up repeatedly in need of sourcing UTXOs from other mixdepths, I think the problem is more deep.

Yes I agree, I think.

I think it's worth delving into the mechanics here a bit, just to get a better sense of what sort of failures we *do* have in sourcing podles. So this is very much sort of background reading material rather than a direct assessment of this PR:

The following parenthetical is background-background, because it *probably* doesn't matter for our purposes:
(
Superficially, if a tumbler run uses very long wait times between transactions , the chance of the in-mixdepth utxos (some or all of which are used in the next transaction) being deep enough in the chain (5 confs) is very high, but vice versa if short wait times.
But it's also an important detail that the "wait before next transaction" clock only starts ticking after the confirmation of the previous, so network tx fees shouldn't affect this (lowball fees can screw up the tumbler timeouts but that's a separate story).
The default tumbler inter-tx wait time *average* is 60 minutes, so 10 confs, which should be more than enough for our 5 ... but it is emphatically an *average* as we all know! So there will definitely be some here and there that are very low comparatively.
)

HOWEVER: in tumbler, if we fail to source commitments due to time, we just wait it out and try again. The trickier part is the 20% size threshold, even if it's a much rarer case:
* If the utxo selection for the join produces 5 or less utxos (should be overwhelming majority), there is no possibility of the first attempt (podle from coins in tx) failing, so that's easy.
* If it selects 6 or more, and none are more than 20% (should be very rare, even here): here's the corner case where we may trigger a fail after this PR that wasn't a fail before. But then the code goes into "tweak schedule mode"; see:

https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/8c57852b2ae3c5a6ca0ec39724495d391982cfc2/jmclient/jmclient/schedule.py#L216-L227

... so at least in the non-sweep case, we trigger a cj amount alteration that will alter utxo selection and that may fix the problem.

* The sweep case is always a lot more likely to hit podle issues I think - by definition it is taking all utxos in the mixdepth, and this might be a large number. More detail: in default tumbler algorithm settings, `-c 2 1` we are are rarely going to see more than 4 or 5 utxos in a mixdepth at any one time during the main tumbler flow; it's far more likely to hit an issue in the starting sweeps that were introduced in 2019/20 (these initial sweeps are described here: https://gist.github.com/chris-belcher/7e92810f07328fdfdef2ce444aad0968 and see #387 - notice that the explicit point of these sweeps is to address the slightly different context of *deposited* coins, as opposed to coins already participating in JM coinjoins) ).
    My assessment of that is these initial sweeps will be absolutely fine (as in, no 20% violation chance) for most use cases where you are *starting with a fresh wallet*. The case that's more difficult (and not only because of podle) is the case of an old heavily used JM wallet being tumbled. Here there's a decent chance of 6+ utxos needing to be swept, and indeed depending on selection algo it can be way higher than 6. There, it's very possible that we fail on the 20% threshold, but counteracting that, for a very heavily used wallet with lots of randomly sized utxos, it'd still be rare for none of them to be much larger than others.

So again, this is all background. I'll do a few tests with carefully "curated" wallet contents on regtest to see if it throws anything up, but I am if anything more convinced than before that this is not really going to be disruptive (and I was already convinced that it was needed!).

Would still like to hear from others though!

chris-belcher commented:
This is a good idea. The code looks very straightforward to me.

AdamISZ commented:
I think you should change the podle sourcing failure message from "these are the utxos in your wallet" to "these are the utxos in mixdepth N:" (and the list of utxos of course); that very detailed output is printed both on CLI and on Qt to help people diagnose the issue when they fail to source podles.

PulpCattel commented:
> If it selects 6 or more, and none are more than 20% (should be very rare, even here): here's the corner case where we may trigger a fail after this PR that wasn't a fail before

IMHO, this is still a good failure. A user has plenty of options to deal with it, like:

* Consolidate few of those UTXOs without CoinJoin and then sweep.
  
* Source another UTXO manually.
  
* Split 1 sweep into 2 sweeps (or whatever needed).

* Let the yield-generator to handle it.

There was a user on Telegram a few weeks ago that had a very similar problem; he wanted to sweep many Wasabi change outputs, but he was unable to source commitments for it.
This to say that this seems a tricky case for PODLEs even without this PR (and hopefully, as you said, this PR doesn't make it much worse in practice).

> I think you should change the podle sourcing failure message from "these are the utxos in your wallet" to "these are the utxos in mixdepth N:" 

Don't you think there is still value in showing the entire wallet there?

My rationale was that, when we trigger this debug messagge, we will almost always follow with a `wallet-tool.py display` or `wallet-tool.py showutxos` to figure out how to procede, so doing it directly in the debug message seems helpful.

I've thought about adding the mixdepth number, but it should be trivial to figure out which is (if not because it's the spending one, for sure by looking at the UTXOs in the message), and adding a `mixdepth` parameter to `generate_podle_error_string` just for that seemed overkill.

> I'm not sure whether we should add a note to that .md about this, or not.

What do you think of 06e0e41, I've added some hints and rephrasing that should make it clear enough(?).
I've not resisted the temptation of splitting some of the super long raw text lines (I can revert it if it's too annoying in the diff), but apart from that it doesn't change much at all.

kristapsk commented:
> The default tumbler inter-tx wait time average is 60 minutes, so 10 confs, which should be more than enough for our 5 ...

It's 6 confs on average, not 10. :)

kristapsk commented:
I think some tests against this should be added to `jmclient/test/test_taker.py:test_make_commitment()`.

AdamISZ commented:
> > The default tumbler inter-tx wait time average is 60 minutes, so 10 confs, which should be more than enough for our 5 ...
> 
> It's 6 confs on average, not 10. :)

Oops, good point :)

AdamISZ commented on docs/SOURCING-COMMITMENTS.md:
```diff
@@ -45,38 +45,41 @@ Be aware that Makers will see this utxo *only if* your usage is successful (they
 1: Utxos that passed age and size limits, but have been used too many times (see taker_utxo_retries in the config):
+dba822a36ad524b775af63db4e25b3558c75ac90d8566a74b2d08b9f63adff97:1
+cde0608bf5426c7ed705d87553718e8c74fa75b9428f14e0759f9e95b03d25f2:1
+2: Utxos that have less than 1 confirmations:
```
It should be 5 not 1?

PulpCattel replied:
Whoops, fixed in f0cd0a9

PulpCattel commented:
WRT improving the test, how would you test this?

Anyway, I'm not familiar with the JM test suite codebase, feel free to improve the test as you want and I'll be happy to run it.

kristapsk commented:
Haven't looked too much at the specific test's code, but idea would be to add bunch of small UTXO's to mixdepth 0, add some big UTXO to mixdepth 1, then try to source commitment for amount for which there is big enough UTXO in mixdepth 1, but not in mixdepth 0, and check that it fails.

AdamISZ commented:
I'll take a look at adding a test case now.

AdamISZ commented:
Pushed some test cases addressing this (in particular cases where utxos outside the mixdepth meet the criteria and those inside do not).
Took more effort because this old test code is a bit messy to say the least, and the existing test of the make_commitment function wasn't really doing much useful (apart from ensuring no crash), but at least functionally this should be fairly thorough now.

I'm ready to merge this shortly. Please report any additional testing results.

kristapsk reviewed (approved):
ACK 4e4b15b166972736ebd4d86de1c72e9002812b42. Test suite passes and code looks correct.

PulpCattel commented:
Test suite passes for me too.

AdamISZ commented:
OK. Tumbler tests also working correctly here.

Since it's only 3 commits (code, docs, test) I don't think we have to squash here, merging.
